From 854a41ed69ffa0dc5ddef7f20af838eadccdd46f Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Sat, 29 Jan 2011 09:30:40 +0000
Subject: [PATCH 150/224] drm/i915/panel: Fix application of legacy backlight value

Remove the spurious division by two in getting the current value, and if
the legacy value is 0xff it is unused. And whilst setting the value, fix
the rounding.

Reported-by: Nick Bowler <nbowler@draconx.ca>
Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=31803
Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
Cc: stable@kernel.org
---
 drivers/gpu/drm/i915/intel_panel.c |   12 ++++++++----
 1 files changed, 8 insertions(+), 4 deletions(-)

--- a/drivers/gpu/drm/i915/intel_panel.c
+++ b/drivers/gpu/drm/i915/intel_panel.c
@@ -173,10 +173,11 @@ u32 intel_panel_get_backlight(struct drm
 		if (is_backlight_combination_mode(dev)){
 			u8 lbpc;
 
-			val &= ~1;
 			pci_read_config_byte(dev->pdev, PCI_LBPC, &lbpc);
+			/* Report the value in [0:max] i.e. always multiplied
+			 * by lbpc even if the hardware does not (0xff).
+			 */
 			val *= lbpc;
-			val >>= 1;
 		}
 	}
 
@@ -205,9 +206,12 @@ void intel_panel_set_backlight(struct dr
 		u32 max = intel_panel_get_max_backlight(dev);
 		u8 lpbc;
 
-		lpbc = level * 0xfe / max + 1;
-		level /= lpbc;
+		lpbc = (level * 0xfe + max-1) / max + 1;
 		pci_write_config_byte(dev->pdev, PCI_LBPC, lpbc);
+
+		/* The hardware is tricksy and treats lpbc == 0xff specially */
+		max = lpbc == 0xff ? 0xff : 0xff * lpbc;
+		level = (level + max-1) / max;
 	}
 
 	tmp = I915_READ(BLC_PWM_CTL);
